### ISSUE TITLE
TypeParser causes StringIndexOutOfBoundsException

### ISSUE THREAD

**Issue by oxisto** [Fraunhofer-AISEC/cpg#153]
The typeparser seems to use some fancy-magic to parse the type string looking for brackets, such as (, [, however, it seems to be missing code that handles if for some reason it cannot find the end-bracket or if multiple brackets are stacked.

Consider the following snippet:

```cpp
#define SOME_MACRO(t) t

void SomeFunction(const char* text=SOME_MACRO(""));
```

The first issue, is that the type parser seems to try to parse the complete function declaration signature to retrieve a type (probably the return type) of it. Not sure what the intent here is.

https://github.com/Fraunhofer-AISEC/cpg/blob/1024f02164d547f961c0f2f3a6d0642410ef6711/src/main/java/de/fraunhofer/aisec/cpg/frontends/cpp/DeclarationListHandler.java#L58-L59

Second issue is, that this will crash within the type parser, since he seems to be confused by the double `)` within the function signature.

```
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 16
	at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:48)
	at java.base/java.lang.String.charAt(String.java:709)
	at de.fraunhofer.aisec.cpg.graph.type.TypeParser.findMatching(TypeParser.java:187)
	at de.fraunhofer.aisec.cpg.graph.type.TypeParser.separate(TypeParser.java:321)
	at de.fraunhofer.aisec.cpg.graph.type.TypeParser.createFrom(TypeParser.java:648)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationListHandler.handleSimpleDeclaration(DeclarationListHandler.java:58)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationListHandler.lambda$new$0(DeclarationListHandler.java:48)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.java:112)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.handleSingleDeclarator(DeclarationHandler.java:199)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.handleSimpleDeclaration(DeclarationHandler.java:171)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.lambda$new$0(DeclarationHandler.java:53)
	at de.fraunhofer.aisec.cpg.frontends.Handler.handle(Handler.java:112)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.handleTranslationUnit(DeclarationHandler.java:261)
	at de.fraunhofer.aisec.cpg.frontends.cpp.CXXLanguageFrontend.parse(CXXLanguageFrontend.java:218)
	at de.fraunhofer.aisec.cpg.TranslationManager.runFrontends(TranslationManager.java:196)
	at de.fraunhofer.aisec.cpg.TranslationManager.lambda$analyze$0(TranslationManager.java:90)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1771)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1763)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

**Comment by vfsrfs** [comment]
It seems that the problem you mention is caused due to multiple different issues.

The purpose of tracking brackets in the TypeParser is to ensure that you are able to parse certain types correctly, since you must ignore certain parts of a definition. For example if you define a two-dimensional array `arr` as in the snippet below you will have to ignore the function call and the other array, but you need to find the other brackets that indicate that the array is two-dimensional. 

```
int arr[foo(otherarr[0])][5];
```

It seems that this is the problem why it is crashing and I will take a look at the bracket-algorithm in order to avoid this problems.


The reason for using the `getRawSignature` is that at the time of developing the TypeSystem, the CDT often generates wrong types e.g. missing pointers, missing arrays, missing qualifiers such as const. This is why the TypeParser was implemented in the first place and why we had to use the rawSignature instead of the values provided by CDT. It seems like the CDT has been updated to 6.10 in the mean time and it might have fixed this problem.

The next issue that is found is regarding the DeclaratorHandler. As you can see in the PR the rawSignature is split by the function name, so that only the return type reaches the TypeParser, and not the entire function. The problem here is, that when requesting the `ctx.getRawSignature` in the DeclarationListHandler the return type disappears, so that only the functionSignature without the return type reaches the typeparser. I don’t know what the reason for this beahavior is, and why the return type disappears.

Additionally when your snippet is modified to

```
#define SOME_MACRO(t) t

void SomeFunction(const char* text=SOME_MACRO(“”)){
}
```

there is no crash. I do not understand why there is a different processing when the function body is defined in comparison to when only the function is defined.

**Comment by oxisto** [comment]
Thanks for looking into this. 

Could you take a look at #154 and see whether this is ok. I have only added a little hack to the bracket parsing, so this probably needs some improvement, but I would appreciate it if you could take my PR as a basis.

Regarding CDT: You need to parse the decl-specifier and the declarator and then CDT gives you the same type as in the raw signature (at least in all test I have done) and this is less-error prone, so that it the way I have implemented it in #154 

Regarding the last one: this is because one is a *declaration* (without function body) and one is a *definition* and they are handled differently in the code. One was using the `getRawSignature`, the other one was doing a hacky `split`-style action to separate the return type. In #154 both are using the new `getTypeStringFromDeclarator` function.

**Comment by vfsrfs** [comment]
I just added a review to the PR #154 . There is an issue regarding function pointers, as they get removed from the raw type string and only the return type is left in the typestring.
